org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/actions/SelectionDispatchAction.java

```diff
@@ -249,7 +251,7 @@ public void run(ISelection selection) {
 
 	@Override
 	public void run() {
-		dispatchRun(getSelection());
+		JavaCore.runReadOnly(() -> dispatchRun(getSelection()));
```

Comment:
Give an example how to do that.

Reply:
* Create a Java project
* Close it
* Close the _Package explorer_ view
* Open the _Package explorer_ view again **but do not select (click) the project**
* Go to the menu _Project > Open_
* Select the project and click _Ok_
* The code walks into `org.eclipse.jdt.ui.actions.OpenProjectAction.internalRun(List<?>)`, line 180

Reply:
In this particular case its not a problem because parameter `boolean fork=true` and the cache is threadlocal.

Reply:
I fail to see the connection here. Can you please explain more in detail:
a. Why is forking safe even though it's a write operation and
b. What could happen with the threadlocal cache if `fork == false`
?

Reply:
I have not tried it but most likely fork=false will result in a major UI freeze. But with this change you should get an IllegalStateException instead.
Forking is ok, because jdt does not care what happens in other threads.